File: occasum-core/src/main/java/pw/stamina/occasum/node/instructions/StandardPropertyNodeInstructionList.java
package pw.stamina.occasum.node.instructions;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.properties.Property;

import java.util.*;

final class StandardPropertyNodeInstructionList implements PropertyNodeInstructionList {
    private final List<PropertyNodeInstruction> instructions;
    private final Map<String, PropertyNodeInstructionList> nestedInstructions;

    StandardPropertyNodeInstructionList() {
        this.instructions = new LinkedList<>();
        this.nestedInstructions = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
    }

    @Override
    public Optional<PropertyNodeInstructionList> findNestedInstructions(String name) {
        return Optional.ofNullable(nestedInstructions.get(name));
    }

    @Override
    public void property(PropertyHandle handle, Property property) {
        //Null handle is permitted
        instructions.add(new PropertyPropertyNodeInstruction(handle, property));
    }

    @Override
    public void property(Property property) {
        property(null, property);
    }

    @Override
    public PropertyNodeInstructionList folder(PropertyHandle handle, String name) {
        if (nestedInstructions.containsKey(name)) {
            instructions.add(new FolderPropertyNodeInstruction(handle, name));
        }

        return nestedInstructions.computeIfAbsent(name, s -> new StandardPropertyNodeInstructionList());
    }

    @Override
    public PropertyNodeInstructionList folder(String name) {
        return folder(null, name);
    }

    @Override
    public Iterator<PropertyNodeInstruction> iterator() {
        return instructions.iterator();
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/instructions/StandardPropertyNodeInstructionList.kt
package pw.stamina.occasum.node.instructions

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.properties.Property

import java.util.*

internal class StandardPropertyNodeInstructionList : PropertyNodeInstructionList {

    private val instructions: MutableList<PropertyNodeInstruction>
    private val nestedInstructions: MutableMap<String, PropertyNodeInstructionList>

    init {
        this.instructions = LinkedList()
        this.nestedInstructions = TreeMap(String.CASE_INSENSITIVE_ORDER)
    }

    override fun findNestedInstructions(name: String): PropertyNodeInstructionList? {
        return nestedInstructions[name]
    }

    override fun property(handle: PropertyHandle, property: Property<*>) {
        instructions.add(PropertyPropertyNodeInstruction(handle, property))
    }

    override fun property(property: Property<*>) {
        instructions.add(PropertyPropertyNodeInstruction(null, property))
    }

    override fun folder(handle: PropertyHandle, name: String): PropertyNodeInstructionList {
        return addFolderInstruction(handle, name)
    }

    override fun folder(name: String): PropertyNodeInstructionList {
        return addFolderInstruction(null, name)
    }

    private fun addFolderInstruction(handle: PropertyHandle?, name: String): PropertyNodeInstructionList {
        if (nestedInstructions.containsKey(name)) {
            instructions.add(FolderPropertyNodeInstruction(handle, name))
        }

        return nestedInstructions.computeIfAbsent(name) { StandardPropertyNodeInstructionList() }
    }

    override fun iterator(): Iterator<PropertyNodeInstruction> {
        return instructions.iterator()
    }
}
